### PR TITLE
fix: zcode empty-string content + codex login cancel barrier

### PR DESCRIPTION
**Pull request by junhoyeo**
## Summary

Two low-severity edge cases surfaced by an independent **Codex** review of the just-merged #769 (Codex login cancellation) and #770 (ZCode parser). Both validated against source before fixing.

### 1. ZCode parser — empty-string content not treated as empty
`crates/tokscale-core/src/sessions/zcode.rs`

`content_chars` normalized `null`, empty array, and empty object to `0` chars — but **not** an empty string. So `content: ""` serialized to `"\"\""` (2 chars) → `estimate_tokens(2) = 1`, causing an assistant entry with empty content to emit a **spurious ~1-token estimated message** instead of taking the zero-token skip path. Added a `Value::String(s) if s.is_empty() => 0` arm.

### 2. Codex login — cancellation isn't a barrier before the persistent import
`crates/tokscale-cli/src/tui/codex_login.rs`

`wait_for_codex_login_child` only observes cancellation **while the child is still running**. If the `codex login` child exited successfully in the same tick the user dismissed the panel, the worker still ran `import_login_auth_file` → `save_account_from_auth_at_path`, **persisting the account to the credentials store despite the cancel** (the UI never showed it because the dropped channel swallowed the event). Now re-checks the `cancelled` flag via a new `login_was_cancelled` helper before the import side effect.

## Severity

Both are narrow edge cases, not regressions in the headline behavior of #769/#770 — Codex reported **no P1**, and no orphaned-process/deadlock in the cancellation handoff.

## Testing

- `cargo fmt --check` ✅
- `cargo clippy --workspace --all-features -- -D warnings` ✅
- `cargo test -p tokscale-core --lib zcode` ✅ 9/9 (added `test_content_chars_treats_empty_string_as_empty`, `test_empty_string_assistant_content_emits_no_message`)
- `cargo test -p tokscale-cli codex_login` ✅ 11/11 (added `login_was_cancelled_tracks_cancellation`)

🤖 Findings identified by Codex; validated and fixed.

---
## Summary by cubic
Fixes two edge cases: the ZCode parser treats empty-string content as empty, and Codex login cancellation is now an atomic barrier to credential persistence. Prevents spurious token estimates and unintended account saves.

- Bug Fixes
  - ZCode: `content_chars` returns 0 for `""`, so empty assistant content no longer emits a fake 1-token message.
  - Codex login: uses `import_unless_cancelled` to hold the child-slot lock across the cancelled check and `auth.json` import, eliminating the check-then-act race so cancelled logins never persist credentials.

<sup>Written for commit 790a24375173a073ebae54580dace3bc0d18da98. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/junhoyeo/tokscale/pull/773?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: crates/tokscale-cli/src/tui/codex_login.rs [hunk 1]
```diff
@@ -121,6 +121,14 @@ fn run_codex_login_in_home(
         anyhow::bail!("{}", codex_login_failure_message(&status, &output_lines));
     }
 
+    // `wait_for_codex_login_child` only observes cancellation while the child
+    // is still running. If the child exited successfully in the same tick the
+    // user dismissed, re-check the flag before importing — otherwise a
+    // cancelled login would still persist the account to the credentials store.
+    if login_was_cancelled(&child_slot) {
```
[the remark this answers is not in the document]
Comment (junhoyeo) [reply crates/tokscale-cli/src/tui/codex_login.rs]: Fixed in `790a2437` (identified by cubic). You're right — the re-check only shrank the window; check-then-import was still a TOCTOU. Replaced `login_was_cancelled` with `import_unless_cancelled`, which holds the slot lock across **both** the cancelled check and `import_login_auth_file`. A concurrent `cancel_codex_login_child` now either runs first (worker sees `cancelled` and bails before importing) or blocks on the lock until the save completes — there is no longer a window where the flag is set but the import still proceeds. Added `import_unless_cancelled_short_circuits_when_cancelled`.
